orchagent/nhgorch.h

```diff
@@ -0,0 +1,242 @@
+    inline const NextHopGroupKey& getKey() const { return m_key; }
+    inline sai_object_id_t getId() const { return m_id; }
+    static inline unsigned int getCount() { return m_count; }
+    inline bool isTemp() const { return m_is_temp; }
+    inline void setTemp(bool is_temp) { m_is_temp = is_temp; }
+    inline bool isSynced() const { return m_id != SAI_NULL_OBJECT_ID; }
+    inline size_t getSize() const { return m_members.size(); }
+
+    /* Convert NHG's details to a string. */
+    std::string to_string() const
+    {
+        return m_key.to_string() + ", SAI ID: " + std::to_string(m_id);
+    }
+
+private:
+
+    /* The next hop group key of this group. */
+    NextHopGroupKey m_key;
+
+    /* The SAI ID of the group. */
+    sai_object_id_t m_id;
+
+    /* Members of this next hop group. */
+    NhgMembers m_members;
+
+    /* Whether the group is temporary or not. */
+    bool m_is_temp;
+
+    /*
+     * Number of existing groups.  Incremented when an object is created and
+     * decremented when an object is destroyed.  This will also account for the
+     * groups created by RouteOrch.
+     */
+    static unsigned int m_count;
+
+    /* Add group's members over the SAI API for the given keys. */
+    bool syncMembers(const std::set<NextHopKey>& nh_keys);
+
+    /* Remove group's members the SAI API from the given keys. */
+    bool desyncMembers(const std::set<NextHopKey>& nh_keys);
+
+    /* Create the attributes vector for a next hop group member. */
+    vector<sai_attribute_t> createNhgmAttrs(
+                                        const NextHopGroupMember& nhgm) const;
+};
+
+/*
+ * Structure describing a next hop group which NhgOrch owns.  Beside having a
+ * unique pointer to that next hop group, we also want to keep a ref count so
+ * NhgOrch knows how many other objects reference the next hop group in order
+ * not to delete them while still being referenced.
+ */
+struct NhgEntry
+{
+    /* Pointer to the next hop group.  NhgOrch is the sole owner of it. */
+    std::unique_ptr<NextHopGroup> nhg;
+
+    /* Number of external objects referencing this next hop group. */
+    unsigned int ref_count;
+
+    NhgEntry() = default;
+    explicit NhgEntry(std::unique_ptr<NextHopGroup>&& _nhg,
+                      unsigned int _ref_count = 0) :
+        nhg(std::move(_nhg)), ref_count(_ref_count) {}
+};
+
+/*
+ * Map indexed by next hop group's CP ID, containing the next hop group for
+ * that ID and the number of objects referencing it.
+ */
+typedef std::unordered_map<std::string, NhgEntry> NhgTable;
+
+/*
+ * Next Hop Group Orchestrator class that handles NEXTHOP_GROUP_TABLE
+ * updates.
+ */
+class NhgOrch : public Orch
+{
+public:
+    /*
+     * Constructor.
+     */
+    NhgOrch(DBConnector *db, string tableName);
+
+    /* Check if the next hop group given by it's index exists. */
+    inline bool hasNhg(const std::string& index) const
+        { return m_syncdNextHopGroups.find(index) !=
```

Comment:
lets keep the function format consistent through out
```
f()
{
     ...
}
```

Reply:
Done